roachtest: sqlsmith/setup=rand-tables/setting=no-mutations failed

[(roachtest).sqlsmith/setup=rand-tables/setting=no-mutations failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1888205&tab=buildLog) on [master@5ad50d887ad95e6b17b0d20fa98bc30319a67c0d](https://github.com/cockroachdb/cockroach/commits/5ad50d887ad95e6b17b0d20fa98bc30319a67c0d):

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/sqlsmith/setup=rand-tables/setting=no-mutations/run_1
	sqlsmith.go:169,sqlsmith.go:199,test_runner.go:753: error: pq: internal error: the first bucket should have NumRange=0
		stmt:
		WITH
			with_266 (col_1696) AS (SELECT * FROM (VALUES (NULL), (9002975591541485424:::INT8)) AS tab_726 (col_1696)),
			with_267 (col_1697)
				AS (
					SELECT
						*
					FROM
						(VALUES (3426907969:::OID), (3773940116:::OID), (0:::OID), (2786855330:::OID)) AS tab_727 (col_1697)
				)
		SELECT
			'05:24:19.890562':::TIME AS col_1698,
			tab_728.col5_11 AS col_1699,
			tab_728.col5_1 AS col_1700,
			tab_728.col5_9 AS col_1701,
			tab_728.col5_2 AS col_1702,
			NULL AS col_1703,
			NULL AS col_1704,
			CASE WHEN tab_728.col5_8 THEN '\x52ea':::BYTES ELSE tab_728.col5_7 END AS col_1705,
			tab_728.col5_0 AS col_1706
		FROM
			defaultdb.public.table5@[0] AS tab_728
		WHERE
			tab_728.col5_8
		ORDER BY
			tab_728.col5_3 ASC
		LIMIT
			13:::INT8;

```

<details><summary>More</summary><p>


Artifacts: [/sqlsmith/setup=rand-tables/setting=no-mutations](https://teamcity.cockroachdb.com/viewLog.html?buildId=1888205&tab=artifacts#/sqlsmith/setup=rand-tables/setting=no-mutations)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Asqlsmith%2Fsetup%3Drand-tables%2Fsetting%3Dno-mutations.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

This looks like an optimizer histogram issue:
```
07:45:10 test_runner.go:615: [w15] ##teamcity[testFailed name='sqlsmith/setup=rand-tables/setting=no-mutations' details='test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/sqlsmith/setup=rand-tables/setting=no-mutations/run_1|n	sqlsmith.go:169,sqlsmith.go:199,test_runner.go:753: error: pq: internal error: the first bucket should have NumRange=0|n		stmt:|n		WITH|n			with_266 (col_1696) AS (SELECT * FROM (VALUES (NULL), (9002975591541485424:::INT8)) AS tab_726 (col_1696)),|n			with_267 (col_1697)|n				AS (|n					SELECT|n						*|n					FROM|n						(VALUES (3426907969:::OID), (3773940116:::OID), (0:::OID), (2786855330:::OID)) AS tab_727 (col_1697)|n				)|n		SELECT|n			|'05:24:19.890562|':::TIME AS col_1698,|n			tab_728.col5_11 AS col_1699,|n			tab_728.col5_1 AS col_1700,|n			tab_728.col5_9 AS col_1701,|n			tab_728.col5_2 AS col_1702,|n			NULL AS col_1703,|n			NULL AS col_1704,|n			CASE WHEN tab_728.col5_8 THEN |'\x52ea|':::BYTES ELSE tab_728.col5_7 END AS col_1705,|n			tab_728.col5_0 AS col_1706|n		FROM|n			defaultdb.public.table5@|[0|] AS tab_728|n		WHERE|n			tab_728.col5_8|n		ORDER BY|n			tab_728.col5_3 ASC|n		LIMIT|n			13:::INT8;|n' flowId='sqlsmith/setup=rand-tables/setting=no-mutations']
07:45:10 test_runner.go:627: [w15] --- FAIL: sqlsmith/setup=rand-tables/setting=no-mutations (84.52s)
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/sqlsmith/setup=rand-tables/setting=no-mutations/run_1
	sqlsmith.go:169,sqlsmith.go:199,test_runner.go:753: error: pq: internal error: the first bucket should have NumRange=0
```

[(roachtest).sqlsmith/setup=rand-tables/setting=no-mutations failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1900812&tab=buildLog) on [master@60c9e055e970bd7f150ebcfad266929b2638d635](https://github.com/cockroachdb/cockroach/commits/60c9e055e970bd7f150ebcfad266929b2638d635):

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/sqlsmith/setup=rand-tables/setting=no-mutations/run_1
	sqlsmith.go:169,sqlsmith.go:199,test_runner.go:753: error: pq: internal error: the first bucket should have NumRange=0
		stmt:
		WITH
			with_187 (col_1074) AS (SELECT NULL AS col_1074),
			with_188 (col_1075) AS (SELECT * FROM (VALUES (0:::OID)) AS tab_458 (col_1075))
		SELECT
			'4d2b:1902:325e:c763:776e:f222:e160:b5dc/18':::INET AS col_1076,
			tab_459.col4_24 AS col_1077,
			0.7824142575263977:::FLOAT8 AS col_1078,
			'4da30a5b-8522-4485-a5d1-5c23ee4e7e6b':::UUID AS col_1079,
			e'kaL\x19uQ%\x1b':::STRING AS col_1080,
			5530919148430863099:::INT8 AS col_1081,
			tab_459.col4_15 AS col_1082,
			tab_459.col4_20 AS col_1083,
			tab_459.col1_13 AS col_1084,
			'1984-10-27':::DATE AS col_1085,
			uuid_v4()::BYTES AS col_1086,
			2321797222:::OID AS col_1087,
			tab_459.col4_28 AS col_1088,
			tab_459.col2_11 AS col_1089,
			(
				SELECT
					'0339edb4-389b-481f-a1de-5292fe3953de':::UUID AS col_1090
				FROM
					defaultdb.public.table1@[0] AS tab_460, defaultdb.public.table1@[0] AS tab_461
				ORDER BY
					tab_460.col1_10 DESC
				LIMIT
					1:::INT8
			)
				AS col_1091,
			tab_459.col4_22 AS col_1092
		FROM
			defaultdb.public.table4@[0] AS tab_459
		WHERE
			tab_459.col2_11
		ORDER BY
			tab_459.col2_11 DESC
		LIMIT
			6:::INT8;

```

<details><summary>More</summary><p>


Artifacts: [/sqlsmith/setup=rand-tables/setting=no-mutations](https://teamcity.cockroachdb.com/viewLog.html?buildId=1900812&tab=artifacts#/sqlsmith/setup=rand-tables/setting=no-mutations)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Asqlsmith%2Fsetup%3Drand-tables%2Fsetting%3Dno-mutations.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/sql/opt/memo/testdata/stats/scan
+++ b/pkg/sql/opt/memo/testdata/stats/scan
@@ -975,3 +975,59 @@ select
  │         └── fd: ()-->(1), (2)-->(4)
  └── filters
       └── z:3 < 990 [type=bool, outer=(3), constraints=(/3: (/NULL - /989]; tight)]
+
+# Regression test for #47742 and #47879. Make sure the first bucket always has
+# NumRange=0, even after filtering.
+exec-ddl
+CREATE TABLE t47742 (a INT, b BOOL, INDEX b_idx (b DESC));
+----
+
+exec-ddl
+ALTER TABLE t47742 INJECT STATISTICS '[
+  {
+    "name":"__auto__",
+    "created_at":"2000-01-01 00:00:00+00:00",
+    "columns":["b"],
+    "row_count":200000,
+    "distinct_count":56128,
+    "null_count":27606,
+    "histo_col_type":"BOOL",
+    "histo_buckets":[{
+      "num_eq":7975541041996628837,
+      "num_range":0,
+      "distinct_range":0,
+      "upper_bound":"false"
+    },
+    {
+      "num_eq":124065620125775458,
+      "num_range":100000000000,
+      "distinct_range":100000000000,
+      "upper_bound":"true"
+    }]
+  }
+]'
+----
+
+opt
+SELECT a, b::string FROM t47742 WHERE b = true
+----
+project
+ ├── columns: a:1(int) b:4(string!null)
+ ├── stats: [rows=2640.64496]
+ ├── fd: ()-->(4)
+ ├── index-join t47742
+ │    ├── columns: a:1(int) t47742.b:2(bool!null)
+ │    ├── stats: [rows=2640.64496, distinct(2)=2.00246926, null(2)=0]
+ │    │   histogram(2)=  0    0    0.0021284 2640.6
+ │    │                <--- false ----------- true
+ │    ├── fd: ()-->(2)
+ │    └── scan t47742@b_idx
+ │         ├── columns: t47742.b:2(bool!null) rowid:3(int!null)
+ │         ├── constraint: /-2/3: [/true - /true]
+ │         ├── stats: [rows=2640.64496, distinct(2)=2.00246926, null(2)=0]
+ │         │   histogram(2)=  0    0    0.0021284 2640.6
+ │         │                <--- false ----------- true
+ │         ├── key: (3)
+ │         └── fd: ()-->(2)
+ └── projections
+      └── t47742.b:2::STRING [as=b:4, type=string, outer=(2)]

--- a/pkg/sql/opt/props/histogram.go
+++ b/pkg/sql/opt/props/histogram.go
@@ -305,10 +305,17 @@ func (h *Histogram) Filter(c *constraint.Constraint) *Histogram {
 	}
 
 	if desc {
-		// Add an empty bucket to indicate that the remaining buckets from the
-		// original histogram have been removed.
+		// After we reverse the buckets below, the last bucket will become the
+		// first bucket. NumRange of the first bucket must be 0, so add an empty
+		// bucket if needed.
 		if iter.next() {
+			// The remaining buckets from the original histogram have been removed.
 			filtered.addEmptyBucket(iter.lb, desc)
+		} else if lastBucket := filtered.buckets[len(filtered.buckets)-1]; lastBucket.NumRange != 0 {
+			iter.setIdx(0)
+			span := makeSpanFromBucket(&iter, prefix)
+			ub := h.getPrevUpperBound(span.EndKey(), span.EndBoundary(), colOffset)
+			filtered.addEmptyBucket(ub, desc)
 		}
 
 		// Reverse the buckets so they are in ascending order.

